### PR TITLE
8271868: Warn user when using mac-sign option with unsigned app-image.

### PR DESCRIPTION
**Pull request by andyherrick**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271868](https://bugs.openjdk.java.net/browse/JDK-8271868): Warn user when using mac-sign option with unsigned app-image.

### Reviewers
 * [Alexander Matveev](https://openjdk.java.net/census#almatvee) (@sashamatveev - **Reviewer**)
 * [Alexey Semenyuk](https://openjdk.java.net/census#asemenyuk) (@alexeysemenyukoracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5004/head:pull/5004` \
`$ git checkout pull/5004`

Update a local copy of the PR: \
`$ git checkout pull/5004` \
`$ git pull https://git.openjdk.java.net/jdk pull/5004/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5004`

View PR using the GUI difftool: \
`$ git pr show -t 5004`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5004.diff">https://git.openjdk.java.net/jdk/pull/5004.diff</a>

</details>

### CODE REVIEW

File: src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java [hunk 1]
```diff
@@ -134,6 +135,18 @@ protected void validateAppImageAndBundeler(
                         I18N.getString(
                             "message.app-image-requires-app-name.advice"));
             }
+            if (SIGN_BUNDLE.fetchFrom(params)) {
+                // if signing bundle with app-image, warn user if app-image
+                // is not allready signed.
+                Path launcher = applicationImage.resolve("Contents/MacOS")
+                        .resolve(APP_NAME.fetchFrom(params));
```
Comment (sashamatveev) [inline src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java:142]: As far as I remember launcher can be signed, but entire app image might not be signed. So, in this case check will pass, but notarization will fail. I think we should run check on app image itself.

File: src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacAppImageBuilder.java [hunk 2]
```diff
@@ -864,4 +864,20 @@ private static String extractBundleIdentifier(Map<String, Object> params) {
...
+      * isFileSigned()
+      * We cannot use this internally to decide whether to unsign a file
+      * or not before resigning, because both unsigned and badly signed
+      * return false. It can be used to differentiate between a signed and
+      * unsigned app-image.
+      */
+     public static boolean isFileSigned(Path file) {
+         try {
+             IOUtils.exec(new ProcessBuilder("/usr/bin/codesign",
+                         "--verify", file.toString()));
```
Comment (sashamatveev) [inline src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacAppImageBuilder.java:877]: --deep should be added, so check is run on nested frameworks and helpers.

**Review by sashamatveev — CHANGES_REQUESTED** [review]

**Comment by alexeysemenyukoracle** [comment]
What if we save "signed/not signed" flag in ".jpackage.xml" when building app image and read its value instead of guessing if they signed app image or not using `codesign`?
.jpackage.xml is designed specifically to pass data between multiple invocations of jpackage.

**Comment by sashamatveev** [comment]
I think it will not work. User might modify app image after it was created and it will invalidate signature.

**Comment by alexeysemenyukoracle** [comment]
What options exist for building a package from app image with invalid signature?

**Comment by andyherrick** [comment]
The purpose of this change is to catch the case (and emit a warning) when user creates an app-image w/o using --mac-sign, and then uses that app image to build a pkg or dmg bundle using --mac-sign.  For this purpose, checking if the main launcher is signed is sufficient.  There is no reason to check all the executables, libraries, and/or Frameworks in the app image., and there is no such thing as signing the app-image itself, also using --deep arg to codesigner is specifically recommended against in all codesigner documentation I have read.
The user can modify the app-image in any way he chooses, possibly invalidating the signing, before using it to create dmg or pkg (or he may create the app-image unsigned, and manually sign all or any part of it.  This really has nothing to do with this change.  The app-image can still be used to create a dmg or pkg using --mac-sign or not.

The alternative of recording if the app-image was created with --app-sign in the AppImageFile is a reasonable alternative to verifying the signing of the main launchers

File: src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java (hunk 1 above)
Comment (andyherrick) [reply src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java:142]: jpackage will either copy the launcher from resource unmodified and unsigned, or sign all the sign-able elements in the app-image (including the launcher).

**Comment by sashamatveev** [comment]
First of all why we only want to cover case when app-image was produced without --mac-sign? If we want to cover such case only, then we need to use approach suggested by Alexey and record signing flag in .jpackage.xml. Otherwise, user will receive false warning if unsigned app-image is provided which was generated without using jpackage or in case if app-image was modified after it was generated and signed.

**Comment by alexeysemenyukoracle** [comment]
My understanding of this enhancement is to warn user when app image created without --mac-sign is used in building a signed package. I.e. to warn user they misused --mac-sign option. We don't want to check the quality of app image and its suitability for packaging in signed package as part of this enhancement.

**Comment by andyherrick** [comment]
now recording in AppImageFile if signing used to create the app-image, and showing warning only if signing an app using an app-image that is not so recorded as signed.

File: src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java [hunk 3]
```diff
@@ -134,6 +136,20 @@ protected void validateAppImageAndBundeler(
                         I18N.getString(
                             "message.app-image-requires-app-name.advice"));
             }
+            if (Optional.ofNullable(
+                    SIGN_BUNDLE.fetchFrom(params)).orElse(Boolean.FALSE)) {
+                // if signing bundle with app-image, warn user if app-image
+                // is not allready signed.
```
Comment (danielpeintner) [inline src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java:142]: nitpicking: typo "allready" -> "already"

**Review by alexeysemenyukoracle — APPROVED** [review]

File: src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java (hunk 3 above)
Comment (andyherrick) [reply src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacBaseInstallerBundler.java:142]: will fix this spelling error

**Review by sashamatveev — APPROVED** [review]

**Pull request closed without merging**